### DISCUSSION TITLE
Can I disable Envoy Metrics and Pilot Metrics in Istio?

### DISCUSSION THREAD

**Discussion by xjtupdy** [istio/istio#48556] [Q&A]
**Clarification Needed: Control and Customization of Istio Envoy Metrics**

I am exploring how Istio manages its metrics, particularly those emitted by Envoy, and I have a couple of questions regarding their configuration and customization.

From my understanding, both proxy-level and service-level metrics in Istio rely on the metrics emitted by Envoy. I'm basing this on the information provided in the [Istio documentation](https://istio.io/latest/docs/ops/configuration/telemetry/envoy-stats/), which states that "Envoy only collects statistical data on items matching the inclusion_list within the stats_matcher JSON element." 

Given this, does Istio have a minimal set of metrics that must always be collected? Additionally, I'm curious if there's any way to completely stop the emission of Envoy metrics in Istio. Is there a specific configuration option for this purpose?

Furthermore, I'm interested in knowing whether it's possible to customize the metrics related to Istio's pilot. Are there any provisions for this kind of customization?

Thank you for any insights or guidance on these points.

**Comment by rootsongjc** [comment]
Istio and Envoy offer a flexible monitoring and telemetry framework. Disabling metrics completely is uncommon and may hinder effective service mesh observation and troubleshooting. Customization focuses on tailoring metric collection and observation rather than stopping it entirely.

**Pilot Metrics Customization**: Customizing Istio's control plane metrics (Pilot) is somewhat limited compared to Envoy metrics. Pilot primarily emits metrics related to configuration and control functions.

**Configuration and Monitoring**: You can monitor these metrics using Prometheus and Grafana, part of Istio's standard telemetry setup. However, customization mainly involves how you monitor and visualize these metrics, not altering what Pilot emits.

I suggest enhancing the [observability best practices](https://istio.io/latest/docs/ops/best-practices/observability/) for further clarity.

**References**

- [Customizing Istio Metrics with Telemetry API](https://istio.io/latest/docs/tasks/observability/metrics/telemetry-api/)

**Reply by xjtupdy** [reply]
Thanks very much for your answer.

In fact, I want to optimize Istio's performance by tailoring metrics emit and collection. So I am trying to test the influence of different metrics on Istio's performance.

It is difficult for me to tailor proxy-level or pilot metrics by only searching info in Istio's doc. 
[Customizing Istio Metrics with Telemetry API](https://istio.io/latest/docs/tasks/observability/metrics/telemetry-api/) can help me exectly cutomize service-level metrics.

Is it appropriate to use benchmark tools below to test the impact of metrics？
https://github.com/istio/tools/tree/master/perf/benchmark

**Reply by rootsongjc** [reply]
Yes, I think you can rely on it.
